src/publish_control.cpp

```diff
@@ -114,6 +133,481 @@ void PublishControl::callback_veh_speed(const pacmod_msgs::VehicleSpeedRpt::Cons
+      turn_signal_cmd_pub.publish(turn_signal_cmd_pub_msg);
+  }
+  else  // Every other controller
+  {
+    if (msg->axes[axes[DPAD_LR]] == AXES_MAX)
+      turn_signal_cmd_pub_msg.command = pacmod_msgs::SystemCmdInt::TURN_LEFT;
+    else if (msg->axes[axes[DPAD_LR]] == AXES_MIN)
+      turn_signal_cmd_pub_msg.command = pacmod_msgs::SystemCmdInt::TURN_RIGHT;
+    else if (msg->axes[axes[DPAD_UD]] == AXES_MIN && msg->buttons[btns[LEFT_BUMPER]] != BUTTON_DOWN)
+      turn_signal_cmd_pub_msg.command = pacmod_msgs::SystemCmdInt::TURN_HAZARDS;
+    else if (local_enable != prev_enable)
+    {
+      if (vehicle_type == VEHICLE_6)
+        turn_signal_cmd_pub_msg.command = pacmod_msgs::SystemCmdInt::TURN_NONE;
+      else
+        turn_signal_cmd_pub_msg.command = last_turn_cmd;
+    }
+    else
+      turn_signal_cmd_pub_msg.command = pacmod_msgs::SystemCmdInt::TURN_NONE;
+
+    if (last_axes.empty() ||
+        last_axes[axes[DPAD_LR]] != msg->axes[axes[DPAD_LR]] ||
+        last_axes[axes[DPAD_UD]] != msg->axes[axes[DPAD_UD]] ||
+        local_enable != prev_enable)
+    {
+      turn_signal_cmd_pub.publish(turn_signal_cmd_pub_msg);
+    }
+  }
+}
+
+void PublishControl::publish_shifting_message(const sensor_msgs::Joy::ConstPtr& msg)
+{
+  // Only shift if brake command is higher than 25%
+  if (last_brake_cmd > 0.25)
+  {
+    pacmod_msgs::SystemCmdInt shift_cmd_pub_msg;
+    shift_cmd_pub_msg.enable = local_enable;
+    shift_cmd_pub_msg.ignore_overrides = false;
+
+    // If the enable flag just went to true, send an override clear and a faults_clear
+    if (!prev_enable && local_enable)
+    {
+      shift_cmd_pub_msg.clear_override = true;
+      shift_cmd_pub_msg.clear_faults = true;
+    }
+
+    uint8_t desired_gear = 0x0;
+    desired_gear |= (msg->buttons[btns[RIGHT_BTN]]  == BUTTON_DOWN) << pacmod_msgs::SystemCmdInt::SHIFT_REVERSE
+                  | (msg->buttons[btns[BOTTOM_BTN]] == BUTTON_DOWN) << pacmod_msgs::SystemCmdInt::SHIFT_HIGH
+                  | (msg->buttons[btns[TOP_BTN]]    == BUTTON_DOWN) << pacmod_msgs::SystemCmdInt::SHIFT_PARK
+                  | (msg->buttons[btns[LEFT_BTN]]   == BUTTON_DOWN) << pacmod_msgs::SystemCmdInt::SHIFT_NEUTRAL;
+
+    switch (desired_gear)
+    {
+      case 1 << pacmod_msgs::SystemCmdInt::SHIFT_REVERSE:
+        shift_cmd_pub_msg.command = pacmod_msgs::SystemCmdInt::SHIFT_REVERSE;
+        break;
+      case 1 << pacmod_msgs::SystemCmdInt::SHIFT_HIGH:
+        shift_cmd_pub_msg.command = pacmod_msgs::SystemCmdInt::SHIFT_HIGH;
+        break;
+      case 1 << pacmod_msgs::SystemCmdInt::SHIFT_PARK:
+        shift_cmd_pub_msg.command = pacmod_msgs::SystemCmdInt::SHIFT_PARK;
+        break;
+      case 1 << pacmod_msgs::SystemCmdInt::SHIFT_NEUTRAL:
+        shift_cmd_pub_msg.command = pacmod_msgs::SystemCmdInt::SHIFT_NEUTRAL;
+        break;
+      // If we've got an invalid command (or multiple buttons pressed) return and don't publish the message
+      default: return;
```

Comment:
Add a ROS_WARN here?

Reply:
If a warning was printed here, then there would be warnings spammed to the console every time any button or stick on the gamepad is used, that isn't related to shifting.